tests/third_party/cupy/test_type_routines.py

```diff
@@ -0,0 +1,94 @@
+import unittest
+
+import numpy
+import pytest
+
+import dpnp as cupy
+from tests.third_party.cupy import testing
+
+
+def _generate_type_routines_input(xp, dtype, obj_type):
+    dtype = numpy.dtype(dtype)
+    if obj_type == "dtype":
+        return dtype
+    if obj_type == "specifier":
+        return str(dtype)
+    if obj_type == "scalar":
+        return dtype.type(3)
+    if obj_type == "array":
+        return xp.zeros(3, dtype=dtype)
+    if obj_type == "primitive":
+        return type(dtype.type(3).tolist())
+    assert False
+
+
+@testing.parameterize(
+    *testing.product(
+        {
+            "obj_type": ["dtype", "specifier", "array", "primitive"],
+        }
+    )
+)
+class TestCanCast(unittest.TestCase):
+    @testing.for_all_dtypes_combination(names=("from_dtype", "to_dtype"))
+    @testing.numpy_cupy_equal()
+    def test_can_cast(self, xp, from_dtype, to_dtype):
+        from_obj = _generate_type_routines_input(xp, from_dtype, self.obj_type)

```

Comment:
I believe it would be better to explicitly skip the parameters value rather than to remove it from the scope (in the future it will be clear why we did so):
```suggestion
            "obj_type": ["dtype", "specifier", "scalar", "array", "primitive"],
        }
    )
)
class TestCanCast(unittest.TestCase):
    @testing.for_all_dtypes_combination(names=("from_dtype", "to_dtype"))
    @testing.numpy_cupy_equal()
    def test_can_cast(self, xp, from_dtype, to_dtype):
        if self.obj_type == "scalar":
            pytest.skip("to be aligned with NEP-50")

        from_obj = _generate_type_routines_input(xp, from_dtype, self.obj_type)

```